src/System.Net.Http/src/System/Net/Http/HttpContent.cs

```diff
@@ -309,14 +305,44 @@ public Task CopyToAsync(Stream stream, TransportContext context)
                 throw new ArgumentNullException("stream");
             }
 
-            TaskCompletionSource<object> tcs = new TaskCompletionSource<object>();
+            TaskCompletionSource<object> tcs = new TaskCompletionSource<object>(this);
             try
             {
                 Task task = null;
                 if (IsBuffered)
                 {
-                    byte[] data = this.GetDataBuffer(_bufferedContent);
-                    task = stream.WriteAsync(data, 0, (int)_bufferedContent.Length);
+                    Debug.Assert(_bufferedContent.Position == 0, "CopyToAsync assumes that _bufferedContent.Position == 0.");
+#if NETNative
+                    // Copy directly out of the MemoryStream buffer
+                    ArraySegment<byte> data;
+                    if (_bufferedContent.TryGetBuffer(out data))
+                        task = stream.WriteAsync(data.Array, data.Offset, data.Count);
+                    else
+                    {
+#endif
+                    task = _bufferedContent.CopyToAsync(stream).ContinueWithStandard(tcs, (copyTask, state) =>
+                    {
+                        // Reset the position of _bufferedContent
+                        var innerTcs = (TaskCompletionSource<object>)state;
+                        var innerThis = (HttpContent)innerTcs.Task.AsyncState;
+                        innerThis._bufferedContent.Position = 0;
+
+                        if (copyTask.IsFaulted)
+                        {
+                            innerTcs.TrySetException(GetStreamCopyException(copyTask.Exception.GetBaseException()));
+                        }
+                        else if (copyTask.IsCanceled)
+                        {
+                            innerTcs.TrySetCanceled();
+                        }
+                        else
+                        {
+                            innerTcs.TrySetResult(null);
```

Comment:
I don't understand why the setting of innerTcs is valid in any of these cases.  ```task``` is now always going to be successfully completed even if the copy operation failed.  Plus this is going to be an unnecessary continuation.  If the whole purpose of this is to reset the position of the stream to 0, why not just add that to the continuation that already exists below?